### PR TITLE
feat: add data aggregation level

### PR DESCRIPTION
**Pull request by nddq**
# Description

This PR introduced the config map option for data aggregation. More details can be found in the docs.
## Related Issue
#138

## Checklist

- [x] I have read the [contributing documentation](https://retina.sh/docs/contributing).
- [x] I signed and signed-off the commits (`git commit -S -s ...`). See [this documentation](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) on signing commits.
- [x] I have correctly attributed the author(s) of the code.
- [x] I have tested the changes locally.
- [x] I have followed the project's style guidelines.
- [x] I have updated the documentation, if necessary.
- [x] I have added tests, if applicable.

## Screenshots (if applicable) or Testing Completed
- Added unit tests for parsing data aggregation level from config file
- Deploy on cluster with different data aggregation level, verified that on `high`, `packetparser` doesn't attach bpf program to eth0
## Additional Notes

Add any additional notes or context about the pull request here.

---

Please refer to the [CONTRIBUTING.md](../CONTRIBUTING.md) file for more information on how to contribute to this project.

### CODE REVIEW

File: controller/Dockerfile.windows-native [hunk 1]
```diff
@@ -4,7 +4,7 @@
 # Maybe one day: https://github.com/moby/buildkit/issues/616
 
 ARG BUILDER_IMAGE
-FROM --platform=windows/amd64 ${BUILDER_IMAGE} as builder
+FROM --platform=windows/amd64 ${BUILDER_IMAGE} AS builder
```
Comment (timraymond) [inline controller/Dockerfile.windows-native:7]: Thanks for fixing all these warnings.

File: pkg/config/config.go [hunk 2]
```diff
@@ -15,16 +15,17 @@ type Server struct {
...
-	LogLevel                 string        `yaml:"logLevel"`
-	EnabledPlugin            []string      `yaml:"enabledPlugin"`
-	MetricsInterval          time.Duration `yaml:"metricsInterval"`
-	EnableTelemetry          bool          `yaml:"enableTelemetry"`
-	EnableRetinaEndpoint     bool          `yaml:"enableRetinaEndpoint"`
-	EnablePodLevel           bool          `yaml:"enablePodLevel"`
-	RemoteContext            bool          `yaml:"remoteContext"`
-	EnableAnnotations        bool          `yaml:"enableAnnotations"`
-	BypassLookupIPOfInterest bool          `yaml:"bypassLookupIPOfInterest"`
+	APIServer                          Server        `yaml:"apiServer"`
```
Comment (timraymond) [inline pkg/config/config.go:18]: Also, thanks for doing `Api` -> `API` :)

File: pkg/config/config.go [hunk 3]
```diff
@@ -15,16 +15,17 @@ type Server struct {
...
+	LogLevel                           string        `yaml:"logLevel"`
+	EnabledPlugin                      []string      `yaml:"enabledPlugin"`
+	MetricsInterval                    time.Duration `yaml:"metricsInterval"`
+	EnableTelemetry                    bool          `yaml:"enableTelemetry"`
+	EnableRetinaEndpoint               bool          `yaml:"enableRetinaEndpoint"`
+	EnablePodLevel                     bool          `yaml:"enablePodLevel"`
+	RemoteContext                      bool          `yaml:"remoteContext"`
+	EnableAnnotations                  bool          `yaml:"enableAnnotations"`
+	BypassLookupIPOfInterest           bool          `yaml:"bypassLookupIPOfInterest"`
+	AttachBPFProgramToDefaultInterface bool          `yaml:"attachBPFProgramToDefaultInterface"`
```
[the remark this answers is not in the document]
Comment (timraymond) [reply pkg/config/config.go:28]: I agree, @anubhabMajumdar, that it will be less confusing ("what BPF program?"). The question is what should the experience be? Should it be an enum like: "aggregationLevel: full" and "aggregationLevel: pod" or something? What's your vision?

**Review by timraymond — COMMENTED** [review]
LGTM on resolving the discussion about the config option UX.

File: pkg/plugin/packetparser/packetparser_linux.go [hunk 4]
```diff
@@ -207,22 +208,28 @@ func (p *packetParser) Start(ctx context.Context) error {
...
-	if err != nil {
-		return err
-	}
-	if len(outgoingLinks) == 0 {
-		return errNoOutgoingLinks
-	}
-	outgoingLink := outgoingLinks[0] // Take first link until multi-link support is implemented
+	if p.cfg.DataAggregationLevel == kcfg.High {
+		p.l.Info("Skipping attaching bpf program to default interface")
+	} else {
```
Comment (anubhabMajumdar) [inline pkg/plugin/packetparser/packetparser_linux.go:213]: Specifically check `DataAggregationLevel == low` for attaching `eth0`. It's very resource consuming to run the eth0 program, so I don't envision it being part of any other level.

File: docs/metrics/dataAggregation.md [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# Data Aggregation
```
Comment (anubhabMajumdar) [inline docs/metrics/dataAggregation.md:1]: Thanks for adding documentation!

File: docs/metrics/dataAggregation.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+
+Retina's data aggregation setting controls the metric volume produced by its agent. A lower level of aggregation yields fewer metrics, and a higher level increases them. The following table details Retina's operational behaviors at each aggregation level.
```
Comment (anubhabMajumdar) [inline docs/metrics/dataAggregation.md:3]: Retina's data aggregation settings allows users to control the number of `Flows` or `events` generated by Retina. When set to high, Retina will serve data about the "most" important flows in a node. This allows Retina to scale to large clusters and be resource efficient. If needed, set aggregation to Low, and you will receive more detailed Flows (ex: packets at different point in the Linux kernel).

File: deploy/legacy/manifests/controller/helm/retina/values.yaml [hunk 7]
```diff
@@ -41,6 +41,7 @@ enablePodLevel: false
 remoteContext: false
 enableAnnotations: false
 bypassLookupIPOfInterest: false
+dataAggregationLevel: 0
```
Comment (snguyen64) [inline deploy/legacy/manifests/controller/helm/retina/values.yaml:44]: i'm not a big fan of having this set as a number because it could be confusing later if we add different levels of data aggregation (0=low, 1=high, 2=med, etc) (not too sure if we are planning to do this). How about we have it as a string value instead of this numerical value? It maps to an enum anyways

File: pkg/plugin/packetparser/packetparser_linux.go [hunk 8]
```diff
@@ -207,22 +208,28 @@ func (p *packetParser) Start(ctx context.Context) error {
...
 
-	outgoingLinks, err := utils.GetDefaultOutgoingLinks()
-	if err != nil {
-		return err
-	}
-	if len(outgoingLinks) == 0 {
-		return errNoOutgoingLinks
-	}
-	outgoingLink := outgoingLinks[0] // Take first link until multi-link support is implemented
+	if p.cfg.DataAggregationLevel == kcfg.High {
```
Comment (snguyen64) [inline pkg/plugin/packetparser/packetparser_linux.go:211]: Question: Would this pretty much disable packetparser? what exactly would packetparser plugin be doing at this level? feels redundant if this is what we're doing when we can just remove this from enabled plugins

**Review by snguyen64 — CHANGES_REQUESTED** [review]

File: pkg/plugin/packetparser/packetparser_linux.go (hunk 8 above)
Comment (nddq) [reply pkg/plugin/packetparser/packetparser_linux.go:211]: we are only disabling attach bpf program to eth0 for high level of aggregation, if pod level metrics is enabled, packetparser would still has to attach bpf programs to veth interfaces associated with the pods.

File: deploy/legacy/manifests/controller/helm/retina/values.yaml (hunk 7 above)
Comment (nddq) [reply deploy/legacy/manifests/controller/helm/retina/values.yaml:44]: yeah that would make more sense, ill switch to that

File: pkg/config/config.go [hunk 9]
```diff
@@ -25,6 +33,7 @@ type Config struct {
 	RemoteContext            bool          `yaml:"remoteContext"`
 	EnableAnnotations        bool          `yaml:"enableAnnotations"`
 	BypassLookupIPOfInterest bool          `yaml:"bypassLookupIPOfInterest"`
+	DataAggregationLevel     Level         `yaml:"dataAggregationLevel"`
```
[the remark this answers is not in the document]
Comment (nddq) [reply pkg/config/config.go:36]: since `Level` is an enum, it should has the value `0`, which equates to `low` by default

File: pkg/config/config.go [hunk 10]
```diff
@@ -46,7 +82,27 @@ func GetConfig(cfgFilename string) (*Config, error) {
 		return nil, fmt.Errorf("fatal error config file: %s", err)
 	}
 	var config Config
-	err = viper.Unmarshal(&config)
+	decoderConfigOption := func(dc *mapstructure.DecoderConfig) {
```
Comment (vakalapa) [inline pkg/config/config.go:85]: What's the default behavior if this value is not set ?

File: pkg/config/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (anubhabMajumdar) [reply pkg/config/config.go]: The current name is looking good - aggregationDataLevel.

File: pkg/config/config.go (hunk 10 above)
Comment (anubhabMajumdar) [inline pkg/config/config.go:85]: This needs detailed comments, especially. the `if` checks.
Also, major candidate to break Retina if `DecodeHook` has bugs or missing edge cases, can you add more unit tests?

File: pkg/plugin/packetparser/packetparser_linux_test.go [hunk 11]
```diff
@@ -36,6 +36,7 @@ var (
 	cfgPodLevelEnabled = &kcfg.Config{
 		EnablePodLevel:           true,
 		BypassLookupIPOfInterest: true,
+		DataAggregationLevel:     kcfg.Low,
```
Comment (anubhabMajumdar) [inline pkg/plugin/packetparser/packetparser_linux_test.go:39]: Unit test for `High` level of data aggregation level?

File: pkg/config/config.go (hunk 10 above)
Comment (nddq) [reply pkg/config/config.go:85]: since Level is an enum, it should have the default value 0, which equates to low

File: docs/metrics/dataAggregation.md [hunk 12]
```diff
@@ -0,0 +1,7 @@
+# Data Aggregation
+
+Retina's data aggregation settings are designed to manage the metric volume produced by its agent. The settings allow users to control the number of Flows or events generated. At a higher aggregation level, fewer metrics are produced, which provides a broad overview and ensures resource efficiency in large clusters. Conversely, a lower level of aggregation results in more metrics being generated, offering detailed information such as packets at different points in the Linux kernel.The operational behaviors of Retina at each aggregation level are detailed in the table below:
```
[the remark this answers is not in the document]
Comment (nddq) [reply docs/metrics/dataAggregation.md:3]: I will add some clarification

File: cmd/legacy/daemon.go [hunk 13]
```diff
@@ -118,6 +118,7 @@ func (d *Daemon) Start() error {
 		zap.String("version", version),
 		zap.String("apiserver", cfg.Host),
 		zap.String("plugins", strings.Join(daemonConfig.EnabledPlugin, `,`)),
+		zap.String("data aggregation level", daemonConfig.DataAggregationLevel.String()),
```
Comment (anubhabMajumdar) [inline cmd/legacy/daemon.go:121]: You have to add this option for `cmd/hubble` as well.

File: pkg/config/config.go [hunk 14]
```diff
@@ -4,18 +4,54 @@ package config
 
 import (
 	"fmt"
+	"reflect"
+	"strings"
 	"time"
 
+	"github.com/mitchellh/mapstructure"
```
Comment (vakalapa) [inline pkg/config/config.go:11]: this is an ARCHIVED repo we cannot use it since there is no support for it going forward. What other alternatives we could use or probably write our own helper function here ?

File: pkg/plugin/packetparser/packetparser_linux.go [hunk 15]
```diff
@@ -207,22 +208,27 @@ func (p *packetParser) Start(ctx context.Context) error {
...
+
+		outgoingLinkAttributes := outgoingLink.Attrs()
+		p.l.Info("Attaching Packetparser",
+			zap.Int("outgoingLink.Index", outgoingLinkAttributes.Index),
+			zap.String("outgoingLink.Name", outgoingLinkAttributes.Name),
+			zap.Stringer("outgoingLink.HardwareAddr", outgoingLinkAttributes.HardwareAddr),
+		)
+		p.createQdiscAndAttach(*outgoingLink.Attrs(), Device)
+	} else {
+		p.l.Info("Skipping attaching bpf program to default interface")
```
Comment (vakalapa) [inline pkg/plugin/packetparser/packetparser_linux.go:230]: mention explicitly that default interface of k8s Node in node ns

File: pkg/config/hubble_config_linux.go [hunk 16]
```diff
@@ -50,6 +50,7 @@ var (
 		EnablePodLevel:           true,
 		LogLevel:                 "info",
 		BypassLookupIPOfInterest: true,
+		DataAggregationLevel:     Low,
```
Comment (vakalapa) [inline pkg/config/hubble_config_linux.go:53]: Can we make default to be high ?

**Review by vakalapa — CHANGES_REQUESTED** [review]

File: pkg/config/config.go (hunk 14 above)
Comment (nddq) [reply pkg/config/config.go:11]: unless `viper` themselves use a different library for decoding, we can't replace this

Comment (vakalapa) [reply pkg/config/config.go:11]: https://github.com/go-viper/mapstructure use this plz

Comment (nddq) [reply pkg/config/config.go:11]: `Unmarshal` is still using the old lib unfortunately 🥲

Comment (nddq) [reply pkg/config/config.go:11]: https://pkg.go.dev/github.com/spf13/viper#DecoderConfigOption

Comment (nddq) [reply pkg/config/config.go:11]: image

Comment (timraymond) [reply pkg/config/config.go:11]: This might be an unpopular opinion, but I feel like our configuration and our CLI is not complicated enough to warrant using either of them.

**Review by anubhabMajumdar — APPROVED** [review]
Please add `Testing Done` before merging.

File: pkg/config/config.go (hunk 14 above)
Comment (nddq) [reply pkg/config/config.go:11]: @timraymond that makes sense, but i _really_ don't want to open that can of worm 🙂

**Comment by nddq** [comment]
> Please add `Testing Done` before merging.

Updated

File: pkg/config/config.go (hunk 14 above)
Comment (timraymond) [reply pkg/config/config.go:11]: @nddq For sure--entirely color commentary on the codebase :)

**Review by vakalapa — APPROVED** [review]
